tests/frontend/test_tdmprogram.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Unit tests for program.py"""
+import pytest
+import numpy as np
+import strawberryfields as sf
+from strawberryfields import ops
+from strawberryfields.tdm import tdmprogram
+from strawberryfields.tdm.tdmprogram import reshape_samples
+
+# make test deterministic
+np.random.seed(42)
+
+
+def test_end_to_end():
+	R =  [1/2, 1, 0.3, 1.4, 0.4] # phase angles
+	BS = [1, 1/3, 1/2, 1, 1/5] # BS angles
+	M =  [1, 2.3, 1.2, 1/2, 5/3] # measurement angles
+	r = 0.8
+	eta = 0.99
+	nbar = 0.1
+	# Instantiate a TDMProgram by passing the number of concurrent modes required
+	# (the number of modes that are alive at the same time).
+	N = 3
+	prog = tdmprogram.TDMProgram(N=N)
+	MM = 1
+	# The context manager has the signature prog.context(*args, copies=1),
+	# where args contains the gate parameters, and copies determines how
+	# many copies of the state should be prepared.
+	# It returns the 'zipped' parameters, and the automatically shifted register.
+	c = 10
+	with prog.context(R, BS, M, copies=c) as (p, q):
+	    ops.Sgate(r, 0) | q[2]
+	    ops.ThermalLossChannel(eta, nbar) | q[2]
+	    ops.BSgate(p[0]) | (q[2], q[1])
+	    ops.Rgate(p[1]) | q[2]
+	    ops.MeasureHomodyne(p[2]) | q[0]
+	eng = sf.Engine('gaussian')
+	result = eng.run(prog)
+	samples = reshape_samples(result.all_samples, c=1)
+	assert samples.shape == (1, len(R)*c)
+
+def test_epr():
+	sq_r = 1.0
```

Comment:
@fab1an-q : This is the test that I was trying to implement. It does not quite work (you can heck this by pulling this branch and running this function).
Once properly implemented the check should be something relatively easy. If we call `var` the variance of the squeezed collective quadrature then one could do something like `assert np.allclose(np.allclose(var, np.exp(-2*r), atol=atol)`

Reply:
I found two errors:
1) The gate parameters are incorrect for producing EPR states. It should be:
```
alpha = [0, np.pi/4]*c
phi = [np.pi/2, 0]*c
```

instead of:
```
alpha = [0, np.pi/2]*c
phi = [np.pi/4, 0]*c
```
I gave you the wrong parameters here, sorry.

2) I think the order in the `with` statement was mixed up:
```
with prog.context(alpha, phi, M, copies=copies) as (p, q):
	ops.Sgate(sq_r, 0) | q[2]
	ops.Rgate(p[0]) | q[2] # the Rgate's argument should be phi, i.e. p[1]
	ops.BSgate(p[1]) | (q[2], q[1]) # the BSgate's argument should be alpha, i.e. p[0]
	ops.MeasureHomodyne(p[2]) | q[0]
```

In the `prog.context` constructor, the order is `alpha, phi`. However, in the `with` statement, `p[0]` corresponds to the `Rgate` (whose argument should be `phi`) and `p[1]` corresponds to the `BSgate` (whose argument should be alpha).

The correct way would be:
```
with prog.context(alpha, phi, M, copies=copies) as (p, q):
    ops.Sgate(sq_r, 0) | q[2]
    ops.BSgate(p[0]) | (q[2], q[1])
    ops.Rgate(p[1]) | q[2]
    ops.MeasureHomodyne(p[2]) | q[0]
```

Reply:
I committed a corrected version including some code to plot nearest-neighbour correlations. This plots now show the characteristic EPR correlations:
![EPR_test](https://user-images.githubusercontent.com/62264296/89881951-6e5fc880-dbc6-11ea-9264-554d374c9121.png)